### PR TITLE
Added Multi Wallet Support for Cardano

### PR DESCRIPTION
**Pull request by Mercurial**
Added Wallets on top of Nami

- Lace
- Eternl
- Flint
- Vespr

Apologies if the implementation is too redundant, probably could have extracted reusable sequences into utility functions.

**Note** right now I noticed theres too many places to edit to add a wallet, perhaps in the future we can improve this and lessen the work needed to add a wallet.

Maybe something similar to how we do it on `cardano-caniuse`: https://github.com/dcSpark/cardano-caniuse/blob/main/src/Wallets.ts

or maybe even auto detection of injected wallets.

https://github.com/ADAPhilippines/Bifrost/blob/203b0a54c2b7ff691274ce45dd2c5747ebe7fb52/src/Bifrost.ts#L34

### CODE REVIEW

**Comment by mkermani144** [comment]
@Mercurial 
Thanks for the PR. Before any code review, we are investigating if we can go on with your implementation, or if we need some refactoring before supporting new wallets. Please wait a bit so we can conclude. It's a high priority nonetheless.

File: apps/rosen/app/(bridge)/BridgeTransaction.tsx [hunk 1]
```diff
@@ -125,19 +130,36 @@ const BridgeTransaction = () => {
...
-              background: 'transparent',
-            }}
-          >
-            <WalletIcon />
-          </Avatar>
-          <Typography>{selectedWallet.label}</Typography>
-        </Grid>
+        <Typography
+          sx={{
+            padding: '5px',
```
Comment (mkermani144) [inline apps/rosen/app/(bridge)/BridgeTransaction.tsx]: We rarely use hard-coded values for spacing. Use `theme.spacing` instead.  
The comment is true for all your other changes, too.

File: apps/rosen/app/(bridge)/ChooseWalletModal.tsx [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+/**
+ * modal to allow user to choose a wallet
+ *
+ * @param open
+ * @param handleClose
+ * @param chainName
+ * @param supportedWallets
+ */
+export const ChooseWalletsModal = ({
```
Comment (mkermani144) [inline apps/rosen/app/(bridge)/ChooseWalletModal.tsx]:
```suggestion
export const ChooseWalletModal = ({
```

File: apps/rosen/app/(bridge)/ChooseWalletModal.tsx [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  Dialog,
+  DialogActions,
+  DialogContent,
+  DialogContentText,
+  Grid,
+  Tooltip,
+} from '@rosen-bridge/ui-kit';
+import { Wallet, WalletInfo } from '@rosen-ui/wallet-api';
+
+interface ChooseWalletsModalProps {
```
Comment (mkermani144) [inline apps/rosen/app/(bridge)/ChooseWalletModal.tsx]:
```suggestion
interface ChooseWalletModalProps {
```

File: apps/rosen/app/(bridge)/ChooseWalletModal.tsx [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+                  item
+                  sx={{
+                    width: '8rem',
+                    height: '8rem',
+                    border: (theme) => `1px solid ${theme.palette.divider}`,
+                    borderRadius: 1,
+                    padding: 2,
+                    margin: 1,
+                    cursor: 'pointer',
+                    display: 'flex', // Added to make the Grid a flex container
```
Comment (mkermani144) [inline apps/rosen/app/(bridge)/ChooseWalletModal.tsx]: Please remove the comments. The code is self-explanatory. In addition, `Grid`s have flex prop themselves and all these flex props are redundant.
The comment is true for all your other changes and need a change.

File: apps/rosen/app/(bridge)/ChooseWalletModal.tsx [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+                      '> *': {
+                        width: '64px',
+                        height: '64px',
+                      },
+                    }}
+                    component="a"
+                    href={link}
+                    target="_blank"
+                  >
+                    <WalletIcon key={label} />
```
Comment (mkermani144) [inline apps/rosen/app/(bridge)/ChooseWalletModal.tsx]: What's the purpose of the `key` prop?

File: wallets/eternl-wallet/README.md [hunk 6]
```diff
@@ -0,0 +1,3 @@
+# @rosen-ui/nami-wallet
```
Comment (mkermani144) [inline wallets/eternl-wallet/README.md]: Please fix these. They occur for other wallets, too.

File: wallets/eternl-wallet/src/connectWallet.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
+/**
+ * handles the nami wallet connection
```
Comment (mkermani144) [inline wallets/eternl-wallet/src/connectWallet.ts]: Same here. Occurs multiple places.

Please check all your changes and make sure no comments, names, etc. mismatch with the current package wallet.

**Review by mkermani144 — CHANGES_REQUESTED** [review]

File: wallets/eternl-wallet/src/connectWallet.ts (hunk 7 above)
Comment (Mercurial) [reply wallets/eternl-wallet/src/connectWallet.ts]: understandable, was assuming this was not too much of a big deal as there was clues to stuff like this already originally: https://github.com/rosen-bridge/ui/blob/3fa552e5c23e366de56cc911ce16dcb9fa57ac8b/wallets/nami-wallet/src/index.ts#L15C16-L15C16

will correct as much as I can though 🙌

**Comment by Mercurial** [comment]
apologies for the quality, this PR was on new years day haha 😅

pushed a commit to address the comments, let me know what else we can do to improve the quality.

File: apps/rosen/app/(bridge)/ChooseWalletModal.tsx [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+                  item
+                  sx={{
+                    width: '8rem',
+                    height: '8rem',
+                    border: (theme) => `1px solid ${theme.palette.divider}`,
+                    borderRadius: 1,
+                    padding: (theme) => theme.spacing(1),
+                    margin: (theme) => theme.spacing(1),
+                    cursor: 'pointer',
+                    flexDirection: 'column',
```
Comment (mkermani144) [inline apps/rosen/app/(bridge)/ChooseWalletModal.tsx]: Use `Grid` component props instead of these CSS props. Please check if the same issue exists in your other changes.

File: wallets/nami-wallet/src/index.ts [hunk 9]
```diff
@@ -12,8 +12,8 @@ export const walletInfo: WalletInfo = {
 };
 
 /**
- * nautilus implementation of the Wallet
- * interface to be able to interact with nautilus wallet
+ * nami implementation of the Wallet
+ * interface to be able to interact with nami wallet
```
Comment (mkermani144) [inline wallets/nami-wallet/src/index.ts]: Though these are not related to your changes, it would be nice to use upper-first version of wallet name (like the new wallets you added):
```suggestion
 * Nami implementation of the Wallet
 * interface to be able to interact with Nami wallet
```

**Review by mkermani144 — CHANGES_REQUESTED** [review]

**Comment by mkermani144** [comment]
Hey @Mercurial, there seems to be a high demand for the new Cardano wallets. It would be nice if you put some time to resolve the other two small comments. If not, we should add new commits after your changes and finish it.

**Comment by Mercurial** [comment]
> Hey @Mercurial, there seems to be a high demand for the new Cardano wallets. It would be nice if you put some time to resolve the other two small comments. If not, we should add new commits after your changes and finish it.

Hello thanks for the additiional review, I removed the additiional `flex` related css now. The component does already use Grid but apologies about mixing grid and flex.

Also updated `nami` casing to match the other wallets!

Huge respect for the quality you guys enforce! 

Let me know if I missed anything else!

File: apps/rosen/app/(bridge)/ChooseWalletModal.tsx [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+                    <WalletIcon />
+                  </Grid>
+                  <Button
+                    onClick={(event) => {
+                      event.preventDefault();
+                      handleConnect(availableWallets[index]);
+                    }}
+                    variant="contained"
+                    size="small"
+                    sx={{ mt: (theme) => theme.spacing(1), width: '100%' }}
```
Comment (mkermani144) [inline apps/rosen/app/(bridge)/ChooseWalletModal.tsx:102]: The button should be disabled if the wallet is not available.
```suggestion
                    sx={{ mt: (theme) => theme.spacing(1), width: '100%' }}
                    disabled={
+                    !availableWallets.find((wallet) => wallet.name === name)
+                  }
```

File: apps/rosen/app/_networks/cardano/index.ts [hunk 11]
```diff
@@ -27,7 +47,13 @@ import { CardanoIcon } from '@rosen-bridge/icons';
 const CardanoNetwork: Network<Wallet> = {
   name: Networks.cardano,
   label: 'Cardano',
-  supportedWallets: [namiWalletInfo],
+  supportedWallets: [
+    namiWalletInfo,
+    laceWalletInfo,
+    eternlWalletInfo,
+    flintWalletInfo,
+    vesprWalletInfo,
```
Comment (mkermani144) [inline apps/rosen/app/_networks/cardano/index.ts]: It seems Vespr is a mobile-only wallet, which is not supported in current implementation of Rosen app. Please remove this line. There is no need to remove other logic related to Vespr, though, as we may support it later.

**Review by mkermani144 — CHANGES_REQUESTED** [review]
Great job, @Mercurial!

During testing your changes, I found the following issues/improvements, too:

The shown balance is not updated when the user switches wallets. This is a new feature that should be added:  

https://github.com/rosen-bridge/ui/blob/c303d7a7341da1280ade2684c5664b47eac685f8/apps/rosen/app/_hooks/useTokenBalance.ts#L41-L49

It should be changed to something like this:
```ts
useEffect(() => {
  const effect = async () => {
    if (
      selectedWallet &&
      tokenField.value &&
      !balanceState.isLoading &&
      (balanceState.token !== token ||
        balanceState.amount !== (await selectedWallet.getBalance(token)))
    )
      getAssetBalance(selectedWallet);
  };

  effect();
}, [selectedWallet, getAssetBalance, tokenField, balanceState, token]);
```

**Comment by Mercurial** [comment]
hey @mkermani144 thanks for the comment. To proceed could I know where I can get an updated tokenMap? That way I could test the balance update feature much more properly. 

Thanks!

I guess I'm not really fully able to test everything out on my side as I've only been using a dummy tokenmap 

```json
{
    "idKeys": {
      "ergo": "tokenId",
      "cardano": "tokenId"
    },
    "tokens": [
      {
        "ergo": {
          "tokenId": "1111111111111111111111111111111111111111111111111111111111111111",
          "name": "test token1",
          "decimals": 0,
          "metaData": {
            "type": "tokenType",
            "residency": "tokenResidency"
          }
        },
        "cardano": {
          "tokenId": "policyId2.assetName2",
          "policyId": "policyId2",
          "assetName": "assetName2",
          "name": "asset1",
          "decimals": 0,
          "metaData": {
            "type": "tokenType",
            "residency": "tokenResidency"
          }
        }
      },
      {
        "ergo": {
          "tokenId": "2222222222222222222222222222222222222222222222222222222222222222",
          "name": "test token2",
          "decimals": 0,
          "metaData": {
            "type": "tokenType",
            "residency": "tokenResidency"
          }
        },
        "binance": {
          "id": "this is a simple ip",
          "name": "binanceAsset",
          "decimals": 0,
          "metaData": {
            "type": "BNBToken",
            "residency": "tokenResidency"
          }
        }
      },
      {
        "ergo": {
          "tokenId": "tokenId",
          "name": "test token3",
          "decimals": 0,
          "metaData": {
            "type": "tokenType",
            "residency": "tokenResidency"
          }
        },
        "cardano": {
          "tokenId": "policyId3.assetName3",
          "policyId": "policyId3",
          "assetName": "assetName3",
          "name": "asset3",
          "decimals": 0,
          "metaData": {
            "type": "tokenType",
            "residency": "tokenResidency"
          }
        }
      }
    ]
  }
  
  ```

**Comment by mkermani144** [comment]
> hey @mkermani144 thanks for the comment. To proceed could I know where I can get an updated tokenMap? That way I could test the balance update feature much more properly.

You can find it in our [contract](https://github.com/rosen-bridge/contract) project, but this is the one currently used in production code:
```json
{
  "idKeys" : {
    "ergo" : "tokenId",
    "cardano" : "tokenId"
  },
  "tokens" : [
    {
      "ergo" : {
        "tokenId" : "e023c5f382b6e96fbd878f6811aac73345489032157ad5affb84aefd4956c297",
        "name" : "rsADA",
        "decimals" : 6,
        "metaData" : {
          "type" : "EIP-004",
          "residency" : "wrapped"
        }
      },
      "cardano" : {
        "tokenId" : "ada",
        "policyId" : "",
        "assetName" : "414441",
        "name" : "ADA",
        "decimals" : 6,
        "metaData" : {
          "type" : "native",
          "residency" : "native"
        }
      }
    },
    {
      "ergo" : {
        "tokenId" : "8b08cdd5449a9592a9e79711d7d79249d7a03c535d17efaee83e216e80a44c4b",
        "name" : "RSN",
        "decimals" : 3,
        "metaData" : {
          "type" : "EIP-004",
          "residency" : "native"
        }
      },
      "cardano" : {
        "tokenId" : "04b95368393c821f180deee8229fbd941baaf9bd748ebcdbf7adbb14.727352534e",
        "policyId" : "04b95368393c821f180deee8229fbd941baaf9bd748ebcdbf7adbb14",
        "assetName" : "727352534e",
        "name" : "rsRSN",
        "decimals" : 3,
        "metaData" : {
          "type" : "CIP26",
          "residency" : "wrapped"
        }
      }
    },
    {
      "ergo" : {
        "tokenId" : "03faf2cb329f2e90d6d23b58d91bbb6c046aa143261cc21f52fbe2824bfcbf04",
        "name" : "SigUSD",
        "decimals" : 2,
        "metaData" : {
          "type" : "EIP-004",
          "residency" : "native"
        }
      },
      "cardano" : {
        "tokenId" : "04b95368393c821f180deee8229fbd941baaf9bd748ebcdbf7adbb14.7273536967555344",
        "policyId" : "04b95368393c821f180deee8229fbd941baaf9bd748ebcdbf7adbb14",
        "assetName" : "7273536967555344",
        "name" : "rsSigUSD",
        "decimals" : 2,
        "metaData" : {
          "type" : "CIP26",
          "residency" : "wrapped"
        }
      }
    },
    {
      "ergo" : {
        "tokenId" : "erg",
        "name" : "ERG",
        "decimals" : 9,
        "metaData" : {
          "type" : "native",
          "residency" : "native"
        }
      },
      "cardano" : {
        "tokenId" : "04b95368393c821f180deee8229fbd941baaf9bd748ebcdbf7adbb14.7273455247",
        "policyId" : "04b95368393c821f180deee8229fbd941baaf9bd748ebcdbf7adbb14",
        "assetName" : "7273455247",
        "name" : "rsERG",
        "decimals" : 9,
        "metaData" : {
          "type" : "CIP26",
          "residency" : "wrapped"
        }
      }
    },
    {
      "ergo" : {
        "tokenId" : "003bd19d0187117f130b62e1bcab0939929ff5c7709f843c5c4dd158949285d0",
        "name" : "SigRSV",
        "decimals" : 0,
        "metaData" : {
          "type" : "EIP-004",
          "residency" : "native"
        }
      },
      "cardano" : {
        "tokenId" : "e6e5dbd69a741e73b0d89d50c23c96dff38a60ab26dcd40077e23704.7273536967525356",
        "policyId" : "e6e5dbd69a741e73b0d89d50c23c96dff38a60ab26dcd40077e23704",
        "assetName" : "7273536967525356",
        "name" : "rsSigRSV",
        "decimals" : 0,
        "metaData" : {
          "type" : "CIP26",
          "residency" : "wrapped"
        }
      }
    },
    {
      "ergo" : {
        "tokenId" : "6ad70cdbf928a2bdd397041a36a5c2490a35beb4d20eabb5666f004b103c7189",
        "name" : "rsHOSKY",
        "decimals" : 0,
        "metaData" : {
          "type" : "EIP-004",
          "residency" : "wrapped"
        }
      },
      "cardano" : {
        "tokenId" : "a0028f350aaabe0545fdcb56b039bfb08e4bb4d8c4d7c3c7d481c235.484f534b59",
        "policyId" : "a0028f350aaabe0545fdcb56b039bfb08e4bb4d8c4d7c3c7d481c235",
        "assetName" : "484f534b59",
        "name" : "HOSKY",
        "decimals" : 0,
        "metaData" : {
          "type" : "CIP26",
          "residency" : "native"
        }
      }
    },
    {
      "ergo" : {
        "tokenId" : "0cd8c9f416e5b1ca9f986a7f10a84191dfb85941619e49e53c0dc30ebf83324b",
        "name" : "COMET",
        "decimals" : 0,
        "metaData" : {
          "type" : "EIP-004",
          "residency" : "native"
        }
      },
      "cardano" : {
        "tokenId" : "d85d4588d4625b370121597ff88f3c1fa66f2a0cb7e4a7a9573a3c3e.7273434f4d4554",
        "policyId" : "d85d4588d4625b370121597ff88f3c1fa66f2a0cb7e4a7a9573a3c3e",
        "assetName" : "7273434f4d4554",
        "name" : "rsCOMET",
        "decimals" : 0,
        "metaData" : {
          "type" : "CIP26",
          "residency" : "wrapped"
        }
      }
    }
  ]
}
```

**Review by mkermani144 — APPROVED** [review]
Thanks for the changes. Your contributions facilitate passages through the bridge!
